xdsl/dialects/transform.py

```diff
@@ -1,41 +1,263 @@
+from __future__ import annotations
+
+from abc import ABC
+from collections.abc import Sequence
+from typing import ClassVar
+
+from attr import field
+
+from xdsl.dialects.builtin import (
+    AnyIntegerAttr,
+    ArrayAttr,
+    DenseArrayBase,
+    DictionaryAttr,
+    StringAttr,
+)
 from xdsl.ir import (
+    Attribute,
     Dialect,
     ParametrizedAttribute,
+    Region,
+    SSAValue,
     TypeAttribute,
 )
 from xdsl.irdl import (
+    AttrSizedOperandSegments,
+    IRDLOperation,
+    ParameterDef,
     irdl_attr_definition,
+    irdl_op_definition,
+    operand_def,
+    opt_attr_def,
+    opt_operand_def,
+    opt_prop_def,
+    prop_def,
+    region_def,
+    result_def,
+    var_operand_def,
+    var_result_def,
 )
+from xdsl.traits import IsolatedFromAbove, IsTerminator
+
+
+class TransformHandleType(ParametrizedAttribute, TypeAttribute, ABC):
+    name: ClassVar[str] = field(init=False, repr=False)
 
 
 @irdl_attr_definition
-class AffineMapType(ParametrizedAttribute, TypeAttribute):
+class AffineMapType(TransformHandleType):
     name = "transform.affine_map"
 
 
 @irdl_attr_definition
-class AnyOpType(ParametrizedAttribute, TypeAttribute):
+class AnyOpType(TransformHandleType):
     name = "transform.any_op"
 
 
 @irdl_attr_definition
-class AnyParamType(ParametrizedAttribute, TypeAttribute):
+class AnyParamType(TransformHandleType):
     name = "transform.any_param"
 
 
 @irdl_attr_definition
-class AnyValueType(ParametrizedAttribute, TypeAttribute):
+class AnyValueType(TransformHandleType):
     name = "transform.any_value"
 
 
+@irdl_attr_definition
+class OperationType(TransformHandleType):
+    name = "transform.op"
+    operation: ParameterDef[StringAttr]
+
+    def __init__(self, operation: str):
+        super().__init__(parameters=[StringAttr(operation)])
+
+
+@irdl_attr_definition
+class ParamType(TransformHandleType):
+    name = "transform.param"
+    type: ParameterDef[TypeAttribute]
+
+    def __init__(self, type: TypeAttribute):
+        super().__init__(parameters=[type])
+
+
+@irdl_attr_definition
+class TypeParamType(TransformHandleType):
+    name = "transform.type"
+
+
+@irdl_attr_definition
+class FailurePropagationModeAttr(ParametrizedAttribute):
+    name = "failures"
+
+
+@irdl_op_definition
+class YieldOp(IRDLOperation):
+    name = "transform.yield"
+
+    traits = frozenset([IsTerminator()])
+
+
+@irdl_op_definition
+class SequenceOp(IRDLOperation):
+    name = "transform.sequence"
+
+    # TODO: Find out how to also use the enum FailurePropagationModeAttr as well as AnyIntegerAttr
+
+    body = region_def("single_block")
+    failure_propagation_mode = prop_def(AnyIntegerAttr)
```

Comment:
why does it need to be anyintegerattr? can it not just be failureproagationmodeattr?

Reply:
both IntegerAttr and FailurePropagationModeAttr are supported. for most use cases, it looks like the IntegerAttr is used so I just used this until FailurePropagationModeAttr was also completed